Report/sections/3_implementation/antlr.tex

```diff
@@ -0,0 +1,57 @@
+
+
+Antlr is written in Java and therefore the Java Runtime Environment must be installed to use it. 
+Once setup, Antlr can convert a grammar into a program that recognizes sentences written in the language defined by the grammar.
+More specifically, Antlr generates a recursive-descent parser that ultimately generates a parse tree.
+This parse tree can then be traversed easily using built-in functionality, and from it we can construct an Abstract Syntax Tree later on.
+As can be seen in \ref{lst:GrammarSnippet}, the grammar syntax recognized by Antlr is the familiar BNF notation.
+
+
+\begin{lstlisting}[caption={A snippet of the \dazel{} grammar used by Antlr to generate the parser}, label={lst:GrammarSnippet}]
+start: gameObject;
+
+gameObject              : gameObjectType=(SCREEN | ENTITY | MOVE_PATTERN) IDENTIFIER gameObjectBlock
+						;
+						
+gameObjectBlock         : L_BRACES gameObjectContents R_BRACES
+						;
+
+empty                   : 
+						;
+
+
+gameObjectContents      : gameObjectContent
+						| gameObjectContent gameObjectContents
+						| empty
+						;
+
+gameObjectContent       : gameObjectContentType=(MAP|ONSCREENENTERED|ENTITIES|EXITS|DATA|PATTERN) statementBlock 
+\end{lstlisting}
+
+In addition, lexer rules can just as easily be defined using regular expressions and strings as can be seen in \ref{lst:LexerSnippet}.
+These rules will also be present within the generated parser such that one may also access and use the lexer rules when constructing the AST. 
+
+\begin{lstlisting}[caption={A snippet of few lexer rules for \dazel{}.}, label={lst:LexerSnippet},escapechar=|]
+SCREEN                  : 'Screen';
+ENTITY                  : 'Entity';
+MOVE_PATTERN            : 'MovePattern';
+MAP                     : 'Map';
+ONSCREENENTERED         : 'OnScreenEntered';
+ENTITIES                : 'Entities';
+EXITS                   : 'Exits';
+DATA                    : 'Data';
+PATTERN                 : 'Pattern';
+
+IDENTIFIER              : [a-zA-Z][a-zA-Z_0-9]*;
+INT                     : [0-9]+;
+FLOAT                   : [0-9]+'.'[0-9]+;	
+\end{lstlisting}
+
+It is important to note that Antlr will look up the lexer rules in the order that they are defined. 
+This means that, if it encounters a sentence structure that is present within multiple different rules, it will attempt to parse it using the first lexer rule, then the second one and so forth until it finds a rule that matches. 
+Because of this, our built-in types are the first rules defined within the lexer as it would otherwise incorrectly parse the keyword Screen using the \texttt{IDENTIFIER} rule instead of the \texttt{SCREEN} rule\cite{parrDefinitiveANTLRReference2012}.
```

Comment:
```suggestion
Because of this, our built-in types are the first rules defined within the lexer as it would otherwise incorrectly parse the keyword Screen using the \texttt{IDENTIFIER} rule instead of the \texttt{SCREEN} rule\cite{parrDefinitiveANTLRReference2012}.
```
Because of this, our built-in types are the first rule defined within the lexer as it would otherwise inccorectly parse keywords such as Screen using the \texttt{IDENTIFIER} rule instead of the \texttt{SCREEN} rule\cite{parrDefinitiveANTLRReference2012}.